### ISSUE TITLE
Align YAML tests and JSON API situation description

### ISSUE THREAD

**Issue by fpagnoux** [openfisca/openfisca-core#716]
Right now, the description of a situation in a YAML test and in the Web API have a slightly different structure:

- In the JSON API, they follow [this structure](http://openfisca.org/doc/openfisca-web-api/input-output-data.html#describing-entities)
- In the YAML tests, the structure in inherited from the former Web API, which is slightly different. 

This is often [confusing for users](https://github.com/openfisca/openfisca-core/issues/599#issuecomment-349316330)

The test runner also currently depends on Scenarios to parse these YAML files, while we've been trying to uncouple scenarios from the core of OpenFisca.  

We should align the test structure to the API one, and process them the same way.

**Comment by fpagnoux** [comment]
Solving this issue means defining a new format for YAML tests. Here are some thoughts:

- As mentioned in the opening post, we want to avoid differences between the YAML test structure and the JSON request one.
    - to avoid confusion
    - so that an user can easily just copy-paste a simulation they sent to the API in a YAML test
- However, strictly limiting YAML tests syntax to the current Web API syntax is really heavy for test writers:
    - Unlike API requests, which are coded "once and for all" and then automatically generated, tests are written manually. Therefore, every piece of boilerplate can become a hassle, and discourage testing. We certainly don't want that!

Trying to balance these two principles, here is a proposition of specification for a new test format:

**Comment by fpagnoux** [comment]
#### RFC 1 - General structure of a YAML test:

```YAML
- name: "Income tax - basic case"
  description: ...
  ...  # other test-level data, such as error margin, reforms to use, etc.
  input:
    ... # Simulation input
  output:
    ...  # Expected output of the simulation
```

The main change relatively to today's format is to:
  - clearly separate input description from test-level data

**Comment by fpagnoux** [comment]
#### RFC 2 - Entityless mode

The following format is valid:

```YAML
- name: "Income tax - basic case"
  input:
    salary:
      2018-10: 2000
  output:
    income_tax:
      2018-10: 300
```

In this case, we make the assumption that:
  - There is only one person in the simulation
  - There is only one instance of each group entity in the simulation (family, household, etc.)
  - This person has, in each group entity, the first role (e.g. `first_parent`, `principal_caregiver`)

**Comment by fpagnoux** [comment]
### RFC 3 - Simplified period mode

The following format is valid:

```YAML
- name: "Income tax - basic case"
  period: 2018-10
  input:
    salary: 2000
  output:
    income_tax: 300
```

In this case, we make the assumption that:
  - All values, input or output, for which no period is explicitly defined, is considered to apply for the global test period.

**Comment by fpagnoux** [comment]
### RFC 3-alt - Simplified period mode (alternative)

The following format is valid:

```YAML
- name: "Income tax - basic case"
  input:
    period: 2018-10
    salary: 2000
  output:
    income_tax: 300
```

In this case, we make the same asumption than in 3, but the period is specified in a different place.
  - Advantage: The "default period" semantically seem to be part of the input
  - Disatvantage: Using a default `period` defined in the input for the output sounds a little funny. We could have a period in each of the two blocks, but that's a little heavy.

**Comment by fpagnoux** [comment]
### RFC 4 - Full entities specification

The following format is valid:

```YAML
- name: "Income tax - couple"
  period: 2017-01
  input:
    persons:
      Alicia:
        salary: 4000
      Javier:
        salary: 2500
    households:
      household_1:
        parents: [Alicia, Javier]
  output:
    persons:
      Alicia:
        income_tax: 600
      Javier:
        income_tax: 375
```

This is _almost_ a Web API-like JSON, except it's using RFC 3 to simplify periods. Of course, fully specifying periods for all input and output would work too.

Note the repetition of `households` and `household_1` for single-household cases, which may be confusing and heavy.

**Comment by fpagnoux** [comment]
### RFC 5 - Full entities specification input, entity-less output

The following format is valid:

```YAML
- name: "Total taxes - couple"
  period: 2017-01
  input:
    persons:
      Alicia:
        salary: 4000
      Javier:
        salary: 2500
    households:
      household_1:
        parents: [Alicia, Javier]
  output:
    total_taxes: 1050
```

When there is no ambiguity, the entity structure doesn't need to be repeated in the output.
  - Note that this works because `total_taxes` applies to household, and there is only one household in the simulation. We could not use it for `income_tax`, which is a person variable, with 2 persons in the simulation.

**Comment by fpagnoux** [comment]
@jvalduvieco, @dlouise64, @Br3nda, @verbman, @guillett, @Morendil, you are the one writing a lot of tests.  Please let us know if:
  - these RFCs seem reasonable to you
  - some of your use-cases are not covered, and you would like to be able to write YAML tests in a different way.

We intend to move on relatively soon to start implementing, but if you think you have an opinion about this topic and need more time to think about it, let us know. 

Note that whatever choices we make here, migrations scripts will be provided to OpenFisca users. 

@jvalduvieco, you wanted to be able to use anchors and references in YAML tests, right? We could try to enable this in the PR to come.

**Comment by guillett** [comment]
Thanks @fpagnoux for make easier the OpenFisca onboarding.

A few comments:
- I am ok with any format as long as there is a script to migrate;
- In case of failing test, please make sure that the entity order is easily accessible (cf. [`entitiesDescription` in `/trace`](https://github.com/openfisca/openfisca-core/blob/master/openfisca_web_api/handlers.py#L64))
- I work with extensions so I would be happy [if I could create tests for a specific extension easily](https://github.com/openfisca/openfisca-core/issues/750).

**Comment by sandcha** [comment]
Thanks @fpagnoux!

For `openfisca-tunisia`, I would be happy to keep the entity less mode (RFC2) and to be able to use a specific extension (in the `description`, like reforms).

Nevertheless, giving one period only might be confusing for someone testing today's taxes on last year incomes.

**Comment by Morendil** [comment]
@fpagnoux I owe you comments on this, but other things took priority, I'm not forgetting it.

**Comment by jvalduvieco** [comment]
Thanks @fpagnoux !
My comments:
* **Anchors & references** - My need is to avoid repeating most of the test case to just change one of the parameters. YAML Anchors and references seem a good way to compose use cases, extends also seems interesting.
*  **RFC 2, RFC 6, RFC 7, RFC 8** - Feel like those are trying to go down the easy path hiding and making assumptions.. For me this is dangerous as the developer needs to know which assumptions are being made or thing on not specified entities, etc... I prefer making things explicit and simple. This combined with YAML tools like references and extends might be enough to have succinct tests.
* **All group entities defined in all use cases** - I understand this is a requirement due internal core implementation but ties all the tests to all group entities defined in the system. This makes tests fragile and is one of my pain points (for example in our system we have `2nd_grade_families`, `parents_and_children_families`, `persons_living_together` group entities. `persons_living_together` only applies to a certain set of formulas but I need to define this for all formulas).
* **All persons in all entities** - This forces me to create `other*` roles to include non relevant persons for a group entity (for example I need to add persons who are not part of  `parents_and_children_families` with a role like `other_persons` to satisfy this requirement).

I guess some of this pains might be avoided with a different design, opinions and guidance are welcome.

**Comment by fpagnoux** [comment]
Thanks a lot @jvalduvieco for your precious inputs! Your 3rd and 4th points will definitely feed discussions on #667, an issue we'll be able to move on soonish, I hope. 

From what I'm reading, and from the few "👎" and "😕"  here and there, there is a pushback against allowing syntaxes that make too much assumptions. I'm thus giving up on RFC 6 and RFC 8.

For 2 and 7 (which have the same minimal level of assumptions), and 5 (which is just a shortcut and doesn't carry on any assumption) to :
-  I'm really concerned that not allowing that we be a huge disincentive to writing tests, especially for economists. 
- YAML anchors and reference can be allowed, but they are not that straightforward, even for developers! 
- A lot of tests in existing models are written that way. Given the constraint we have on specifying all entities, I'm really not sure turning 5-lines entity-less tests into 20-lines fully specified tests would bring much clarity.

**Comment by Morendil** [comment]
@fpagnoux Belatedly…

One thing I would like to have here is *conceptual clarity*, taking priority over *syntactic convenience*. So for instance I like **RFC 1** because it conveys what a test (and by extension an OpenFisca transaction) is - we treat OF as a stateless "black box" that will take inputs and yield outputs.

I find **RFC 2** potentially confusing because it raises questions I have no idea how to answer:
- what happens when an input or an output variable is used that belongs to a non-Person entity?
- what if there are two variables with the same name attached to distinct entities?

The same applies to a lesser degree to **RFC 5**.

Why is the Person entity given such primacy in the first place? What if I'd like to I write a test that focuses on a Family variable only in the style of **RFC 2**, for instance to check the [PAJE formula](https://github.com/openfisca/openfisca-france/blob/cb8c0c189846fdc053a7448d8ba22f476cff763b/openfisca_france/model/prestations/prestations_familiales/paje.py#L62)?

Ideally I could write something like
```YAML
- name: "PAJE - can we add"
  entities: familles
  input:
    paje_base:
      2018-10: 1000
    paje_naissance:
      2018-10: 1000
  output:
    paje:
      2018-10: 2000
```

And yes, this test might require me to specify additional input variables because the first calculation that navigates from the Famille entity to a Person in that Famille will crash, but that's my problem as the test writer. The general principle is that **writing a test should require the least possible amount of background knowledge on what the test runner will be doing with the data**. If any assumption is made (for instance that Person is the default entity) it should be possible to override the default, in a documented manner.

Another thing I don't understand is the nature of the constraint "Every person has to belong to one of each group entity (e.g. household)". From where does that arise? Can it be relaxed? Can we just let the test crash if there are entities that do not have a person assigned, and we try to navigate from a Person to their Family? Or perhaps better, allow the inferences in **RFC 2** as a default, which can be overriden with the **RFC 4** syntax.

It seems to me that this aligns also with "do the least amount of work possible in translating test data syntax into a runnable simulation".

Periods are a somewhat lesser source of confusion although #564 shows that "setting inputs for a given period" can be non-trivial.

The bottom line here is that IMO the syntax is a second-order thing, compared to having a reliable and simple mental model of what happens when I specify a test; and I think the overall effort is on the right track in ditching the scenarios approach.

**Comment by guillett** [comment]
An uncovered use case:

Easily change the period of a test.

Given that  `period` is specified it is somehow duplication to indicate for instance:
```yaml
- salary:
    2018-01: 1001
    2018-02: 1002
```
This could be replaced with

```yaml
- salary:
    M-1: 1001
    M-2: 1002
```

**Comment by fpagnoux** [comment]
> what happens when an input or an output variable is used that belongs to a non-Person entity?

Maybe one element that is was not 100% clear: I took an example with a person variable, but it would work the same with a family variable:


```YAML
- name: "Total taxes include housing tax"
  input:
    housing_tax:
      2018: 1000
  output:
    total_taxes:
      2018: 1000
```

> what if there are two variables with the same name attached to distinct entities?

That is not allowed in OpenFisca. Variable names must be unique, even for different entities.

**Comment by Morendil** [comment]
> Variable names must be unique, even for different entities.

In that case it's not unreasonable to ask "why would you ever want to specify entities in tests"? The only answer I can come up with is that otherwise you might lose track of the length of your input vectors, but since so few of our tests ever contain more than one of anything…

**Comment by fpagnoux** [comment]
> In that case it's not unreasonable to ask "why would you ever want to specify entities in tests"? The only answer I can come up with is that otherwise you might lose track of the length of your input vectors, but since so few of our tests ever contain more than one of anything…

That's why, at least for France, the huge majority of tests do not specify entities (and why I'm really reluctant to prohibit this).

There are cases where specifying the entities becomes necessary, for instance to test the impact of a child's income on a family benefit.

**Comment by Morendil** [comment]
It's not necessarily clear to everyone (it's not to me) what conclusions we've come to after this RFC, and from looking at [WIP](https://github.com/openfisca/openfisca-core/tree/uniformize-tests-api) some of these options are being implemented which will make them _de facto_ decisions, so I'd suggest a recap to make it clear where we are heading.

Is there really a lot of value from changing `{input,output}_variables` to the shorter `{input,output}`?

What I'm seeing by looking at WIP is that this is, so far, the _only_ change to some test files. But if many existing test files are already in conformity with the new format, and only some of the more complex ones need to be changed, I would strongly favor backward-compatible naming of the keys. This way large models like France are spared a large migration to the new format.

**Comment by fpagnoux** [comment]
> This way large models like France are spared a large migration to the new format.

I'm don't think we can spare the large migration to France, and this renaming is not the hard part 😓 .

> Is there really a lot of value from changing {input,output}_variables to the shorter {input,output}?

Not a huge one, but it makes it really easy to check if a test has already been migrated, which has some advantages for the migration script.

> I'd suggest a recap to make it clear where we are heading

- RFC 8 and 3alt are dropped
- The others will be implemented, as they are _massively_ used it current tests in France, and dropping them would make the migration much more complicated

**Comment by Br3nda** [comment]
Just fyi, we're hoping to do a visualisation of the tests, from the yaml, to help explain the concept to non coders. We'll delay until the new format is somewhat stabilised into core.

**Comment by Br3nda** [comment]
Catching up on this thread, we use entities in many tests in Aotearoa (New Zealand's) openfisca rules. We have entitlements that an adult gets if they have, for example, a disabled child in their care. We also have some entitlements to a rebate for a property (a home) based on the occupants. Many other examples. Much of the law we've encoded so far is all about the relationships between entities.

**Comment by Br3nda** [comment]
Gah hit close button by mistake

**Comment by fpagnoux** [comment]
Done 🎉
